extend benchmarks

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54652 Add runner for instruction count benchmarks.
* **#54651 extend benchmarks**

This PR fleshes out the benchmarks to everything I could come up with. (166 individual cases when all is said and done.) If there's anything you feel warrants a spot in CI that I've missed, by all means let me know.

Differential Revision: [D27537824](https://our.internmc.facebook.com/intern/diff/D27537824)

Commit:
extend benchmarks

[ghstack-poisoned]

bdhirsh commented on benchmarks/instruction_counts/definitions/standard.py:
```diff
@@ -51,17 +51,79 @@
+            x += y_float                             | x += y_float;
+            x += y_vector                            | x += y_vector;
+            x += y_int                               | x += y_int;
+            x + y_float                              | x + y_float;
```
Is it worth adding separate `out` variant benchmarks, e.g. `torch.add(y, y, out=x)`, `torch::add_out(x, y, y);`? I had a recent PR that only improved `out` variants and not functional/inplace, so I ended up adding an `add_out` benchmark locally to this suite to measure: https://github.com/pytorch/pytorch/pull/53575.

It might be worth measuring `out` variants more explicitly if we know that they're used frequently (e.g. in static runtime).

I also don't want to over-index on that PR though, since most of the code-paths in the functional/out variants are the same. It probably depends on how comprehensive we want the whole suit to be.

robieta replied:
It's a good point. I had kind of treated `x += y` as the same as `torch.add(x, y, out=x)`, but they're really not the same. (Or at least they don't have to be.) I've added both in-place and out of place versions of `torch.` / `torch::` for add and mul.

ezyang commented on benchmarks/instruction_counts/definitions/standard.py:
```diff
@@ -51,17 +51,79 @@
-        },
+        "Math": GroupedVariants(*parse_stmts(r"""
+            Python                                   | C++
+            ---------------------------------------- | ----------------------------------------
```
It seems like it'll be pretty difficult to tell if you have sufficient coverage of framework code with the benchmarks. Which kind of suggests some sort of way of running coverage on the performance tests might be a good idea.

robieta replied:
Can we defer this to a follow up? I think we need to put some thought into exactly what a useful definition of "coverage" would be.

ezyang replied:
Yes, please defer

ezyang commented:
That's a bit surprising, and I'm not sure why it doesn't work. Guess we need to file a bug. Are you linking against ATen_cpu in this setup?

ngimel commented on benchmarks/instruction_counts/definitions/standard.py:
```diff
@@ -51,17 +51,79 @@
+            x.contiguous()                           | x.contiguous();
+                                                     |
+            # @contiguous (non-trivial)              | // @contiguous (non-trivial)
+            x.t().contiguous()                       | x.t().contiguous();
```
This benchmarks `.t().contiguous()`, might be better to separate`.t()` into setup.

robieta replied:
Good point. I moved the transpose to setup, and also added a transpose benchmark. (Even though it's a degenerate case of `as_strided` it's a super important / common one.)

ngimel commented on benchmarks/instruction_counts/definitions/standard.py:
```diff
@@ -97,14 +159,57 @@
+        x.unsqueeze(0)                           | x.unsqueeze(0);
+                                                 |
+        # @view                                  | // @view
+        x.view(-1, 1)                            | x.view({-1, 1});
```
can you please add `.reshape` here? It's much more expensive than `view` now, would be good to track and fix it

robieta replied:
Done. Also, thanks for filing https://github.com/pytorch/pytorch/issues/55126

swolchok commented:
> If there's anything you feel warrants a spot in CI that I've missed, by all means let me know.

I know there may be some logistical difficulties, but perhaps it would be nice to sometimes run these on CUDA as well? (I ran `rg -i cuda` under pytorch/benchmarks and didn't see anything, and did see a reference to `kCPU`; sorry if I just missed this)


More generally, should we get this in and add more benchmarks as we go? Doesn't seem like we have to gate increasing coverage on *fully* increasing coverage.

Commit:
Update on "extend benchmarks"


This PR fleshes out the benchmarks to everything I could come up with. (166 individual cases when all is said and done.) If there's anything you feel warrants a spot in CI that I've missed, by all means let me know.

[ghstack-poisoned]

robieta commented:
> perhaps it would be nice to sometimes run these on CUDA as well?

You are correct, there are no CUDA benchmarks. (@ngimel had the same request, so there's clearly quorum.) The main thing I think will take some thought is that Callgrind is benchmarking CPU only, so we would be measuring launch rather than compute. Which is still important, I just am not familiar enough with the cuda part of the runtime to know how to best compose those overhead benchmarks.

In general I don't think we should gate this PR on cuda, since there will be other runner changes (like `CUDA_VISIBLE_DEVICES`, and only running if a GPU is present) that would have to go in as well. But definitely agree it should be a follow up.

ezyang commented:
callgrind overhead is still fine, because cpu overhead = less gpu utilization. It won't tell you anything about how fast the cuda kernels themselves are running, but you can nail that some other way. One question is how deterministic going through the CUDA runtime is; there's definitely multithreading involved underneath.

Commit:
Update on "extend benchmarks"


This PR fleshes out the benchmarks to everything I could come up with. (166 individual cases when all is said and done.) If there's anything you feel warrants a spot in CI that I've missed, by all means let me know.

[ghstack-poisoned]

Commit:
Update on "extend benchmarks"


This PR fleshes out the benchmarks to everything I could come up with. (166 individual cases when all is said and done.) If there's anything you feel warrants a spot in CI that I've missed, by all means let me know.

Differential Revision: [D27537824](https://our.internmc.facebook.com/intern/diff/D27537824)

[ghstack-poisoned]
